src/coprocessor/dag/batch/executors/top_n_executor.rs

```diff
@@ -0,0 +1,718 @@
+        if descriptor.get_order_by().is_empty() {
+            return Err(box_err!("Missing Top N column"));
+        }
+        for item in descriptor.get_order_by() {
+            RpnExpressionBuilder::check_expr_tree_supported(item.get_expr())?;
+        }
+        Ok(())
+    }
+}
+
+impl<Src: BatchExecutor> BatchTopNExecutor<Src> {
+    #[cfg(test)]
+    pub fn new_for_test(
+        src: Src,
+        order_exprs: Vec<RpnExpression>,
+        order_is_desc: Vec<bool>,
+        n: usize,
+    ) -> Self {
+        assert_eq!(order_exprs.len(), order_is_desc.len());
+
+        Self {
+            heap: BinaryHeap::new(),
+            eval_columns_buffer_unsafe: Box::new(Vec::new()),
+            order_exprs: order_exprs.into_boxed_slice(),
+            order_is_desc: order_is_desc.into_boxed_slice(),
+            n,
+
+            context: EvalContext::default(),
+            src,
+            is_ended: false,
+        }
+    }
+}
+
+impl<Src: BatchExecutor> BatchTopNExecutor<Src> {
+    pub fn new(
+        config: Arc<EvalConfig>,
+        src: Src,
+        order_exprs_def: Vec<Expr>,
+        order_is_desc: Vec<bool>,
+        n: usize,
+    ) -> Result<Self> {
+        assert_eq!(order_exprs_def.len(), order_is_desc.len());
+
+        let mut order_exprs = Vec::with_capacity(order_exprs_def.len());
+        for def in order_exprs_def {
+            order_exprs.push(RpnExpressionBuilder::build_from_expr_tree(
+                def,
+                &config.tz,
+                src.schema().len(),
+            )?);
+        }
+
+        Ok(Self {
+            // Avoid large N causing OOM
+            heap: BinaryHeap::with_capacity(n.min(1024)),
+            // Simply large enough to avoid repeated allocations
+            eval_columns_buffer_unsafe: Box::new(Vec::with_capacity(512)),
+            order_exprs: order_exprs.into_boxed_slice(),
+            order_is_desc: order_is_desc.into_boxed_slice(),
+            n,
+
+            context: EvalContext::new(config),
+            src,
+            is_ended: false,
+        })
+    }
+
+    #[inline]
+    fn handle_next_batch(&mut self) -> Result<Option<LazyBatchColumnVec>> {
+        // Use max batch size from the beginning because top N
+        // always needs to calculate over all data.
+        let src_result = self
+            .src
+            .next_batch(crate::coprocessor::dag::batch_handler::BATCH_MAX_SIZE);
+
+        self.context.warnings = src_result.warnings;
+
+        let src_is_drained = src_result.is_drained?;
+
+        if src_result.data.rows_len() > 0 {
+            self.process_batch_input(src_result.data)?;
+        }
+
+        if src_is_drained {
+            Ok(Some(self.heap_take_all()))
+        } else {
+            Ok(None)
+        }
+    }
+
+    #[allow(clippy::transmute_ptr_to_ptr)]
+    fn process_batch_input(&mut self, mut data: LazyBatchColumnVec) -> Result<()> {
+        let src_schema_unbounded = unsafe { &*(self.src.schema() as *const _) };
```

Comment:
I think to use `std::mem::transmute(&self.src.schema())` to extend explicitly lifetime is more readable. e.g:
```rust
let src_schema_unbounded: &'static ... = std::mem::transmute(&self.src.schema())
```
Everyone can understand the meaning immediately.

Reply:
- -| Looks like we diverge in this. Shall we add an unsafe util function with name "extend_lifetime"?

Reply:
@sticnarf I think it's better to add this function.